core/src/main/java/org/jdbi/v3/core/argument/AbstractArgumentFactory.java

```diff
@@ -95,12 +95,16 @@ public final Optional<Argument> build(Type type, Object value, ConfigRegistry co
         if (!isInstance.test(type, value)) {
             return Optional.empty();
         }
-        return innerBuild(value, config);
+        Optional<Argument> arg = innerBuild(value, config);
+        if (!arg.isPresent()) {
+            throw new UnableToCreateStatementException("Prepared argument " + value + " of type " + type + " failed to build");
+        }
+        return arg;
     }
 
     @SuppressWarnings("unchecked")
     private Optional<Argument> innerBuild(Object value, ConfigRegistry config) {
-        return Optional.of(value == null
+        return Optional.ofNullable(value == null
```

Comment:
wouldn't

```java
...
    Argument arg = (value == null ? new NullArgument(sqlType) : build ((T) value, config));
    if (arg == null) {
        throw new UnableToCreateStatementException("Prepared argument " + value + " of type " + type + " failed to build");
    }
   return Optional.of(arg);
}
```

and leave the other part unchanged be simpler? The innerBuild is called in multiple places and the check above only covers one place? Worst case, you need to pass the type argument into innerBuild to make the error message work.